substrate/frame/revive/src/wasm/runtime.rs

```diff
@@ -2152,4 +2152,44 @@ pub mod env {
 			already_charged,
 		)?)
 	}
+
+	/// Sets the storage at a fixed 256-bit key with a fixed 256-bit value.
+	/// See [`pallet_revive_uapi::HostFn::set_storage_or_clear`].
+	#[stable]
+	#[mutating]
+	fn set_storage_or_clear(
+		&mut self,
+		memory: &mut M,
+		flags: u32,
+		key_ptr: u32,
+		value_ptr: u32,
+	) -> Result<u32, TrapReason> {
+		let value = memory.read_h256(value_ptr)?;
+		if value.as_bytes().iter().all(|&b| b == 0) {
+			self.clear_storage(memory, flags, key_ptr, SENTINEL)
+		} else {
+			self.set_storage(memory, flags, key_ptr, SENTINEL, value_ptr, 32)
```

Comment:
We should avoid reading the value twice. It can be non free when we move PolkaVM into the client. I suggest changing the signature of `set_storage` to be able to accept both `ptr + len` or the value itself.